extension/content/utils/authSession.js

```diff
@@ -0,0 +1,263 @@
+import auth0 from "auth0-js";
+import autobind from "autobind-decorator";
+
+const LOGIN_FAILED_CODES = [
+  "login_required",
+  "consent_required",
+  "interaction_required",
+];
+
+@autobind
+class AuthSession {
+  constructor(key, environment) {
+    this.key = key;
+
+    this.handlers = {
+      login: [],
+      logout: [],
+    };
+
+    this.webAuth = new auth0.WebAuth({
+      domain: environment.oidcDomain,
+      audience: `https://${environment.oidcDomain}/userinfo`,
+      clientID: environment.oidcClientId,
+      redirectUri: browser.identity.getRedirectURL(),
+      responseType: "token id_token",
+      scope: "openid profile email",
+    });
+  }
+
+  get expiresAt() {
+    return JSON.parse(
+      localStorage.getItem(`environment.${this.key}.auth.expiresAt`),
+    );
+  }
+
+  get authResult() {
+    return JSON.parse(
+      localStorage.getItem(`environment.${this.key}.auth.result`),
+    );
+  }
+
+  get accessToken() {
+    return this.authResult && this.authResult.accessToken;
+  }
+
+  get profile() {
+    return this.authResult && this.authResult.idTokenPayload;
+  }
+
+  /**
+   * Register a new event handler.
+   *
+   * @param {string} event  The name of the event.
+   * @param {function} handler  A callback function.
+   */
+  registerHandler(event, handler) {
+    if (!this.handlers[event]) {
+      throw new Error("Invalid event.");
+    }
+    this.handlers[event].push(handler);
+  }
+
+  /**
+   * Unregister an event handler.
+   *
+   * @param {string} event  The name of the event.
+   * @param {function} handler  A callback function.
+   */
+  unregisterHandler(event, handler) {
+    if (!this.handlers[event]) {
+      throw new Error("Invalid event.");
+    }
+    this.handlers[event] = this.handlers[event].filter(h => h !== handler);
+  }
+
+  /**
+   * Generate a random string to use as a nonce.
+   *
+   * @param {int} length  The length of the generated string.
+   * @returns {string}
+   */
+  generateNonce(length) {
```

Comment:
I have a bunch of questions about this. Mostly, where did it come from? Should we be "improving" it, or should we leave it as is because it came from someone like Duo? If so, we should probably leave a comment to that effect.

Reply:
I'll add a comment but it came from Auth0's docs: https://auth0.com/docs/api-auth/tutorials/nonce#generate-a-cryptographically-random-nonce

They also mention using [nanoid](https://github.com/ai/nanoid) so that's another option.